src/rdkafka.c

```diff
@@ -270,7 +279,463 @@ void rd_kafka_log0 (const rd_kafka_conf_t *conf,
         rd_kafka_log_buf(conf, rk, level, fac, buf);
 }
 
+static int check_oauthbearer_extension_key(const char *key,
+                char *errstr, size_t errstr_size) {
+        if (!strcmp(key, "auth")) {
+		rd_snprintf(errstr, errstr_size,
+		        "Cannot explicitly set the reserved `auth` SASL/OAUTHBEARER extension key");
+                return -1;
+        }
+        /*
+         * https://tools.ietf.org/html/rfc7628#section-3.1
+         * key            = 1*(ALPHA)
+         * 
+         * https://tools.ietf.org/html/rfc5234#appendix-B.1
+         * ALPHA          =  %x41-5A / %x61-7A   ; A-Z / a-z
+         */
+        if (*key == '\0') {
+		rd_snprintf(errstr, errstr_size,
+		        "SASL/OAUTHBEARER extension keys must not be empty");
+                return -1;
+        }
+        const char *c = key;
+        while (*c != '\0') {
+                if (!(*c >= 'A' && *c <= 'Z') && !(*c >= 'a' && *c <= 'z')) {
+                        rd_snprintf(errstr, errstr_size,
+                                "SASL/OAUTHBEARER extension keys must only consist of A-Z or a-z characters: %s (%c)",
+                                key, *c);
+                        return -1;
+                }
+                ++c;
+        }
+        return 0;
+}
+
+static int check_oauthbearer_extension_value(const char *value,
+                char *errstr, size_t errstr_size) {
+        /*
+         * https://tools.ietf.org/html/rfc7628#section-3.1
+         * value          = *(VCHAR / SP / HTAB / CR / LF )
+         * 
+         * https://tools.ietf.org/html/rfc5234#appendix-B.1
+         * VCHAR          =  %x21-7E  ; visible (printing) characters
+         * SP             =  %x20     ; space
+         * HTAB           =  %x09     ; horizontal tab
+         * CR             =  %x0D     ; carriage return
+         * LF             =  %x0A     ; linefeed
+         */
+        const char *c = value;
+        while (*c != '\0') {
+                if (!(*c >= '\x21' && *c <= '\x7E') && *c != '\x20' && *c != '\x09' && *c != '\x0D' && *c != '\x0A') {
+                        rd_snprintf(errstr, errstr_size,
+                                "SASL/OAUTHBEARER extension values must only consist of space, horizontal tab, CR, LF, and visible characters (%%x21-7E): %s (%c)",
+                                value, *c);
+                        return -1;
+                }
+                ++c;
+        }
+        return 0;
+}
+
+int rd_kafka_oauthbearer_set_token(rd_kafka_t *rk,
+                const char *token_value, int64_t md_lifetime_ms,
+                const char *md_principal_name,
+                const char **extensions, size_t extension_size) {
+        /* Check args for correct format */
+        if (!token_value) {
+		rd_kafka_set_fatal_error(rk, RD_KAFKA_RESP_ERR__INVALID_ARG,
```

Comment:
Fatal errors are only to be used when the client can't continue, such as when the idempotent producer guarantees can't be fulfilled.

Invalid input is a user error that the client doesn't care about, it is up to the application to handle it.
Instead have it return RD_KAFKA_RESP_ERR__INVALID_ARG.

But you typically don't need to check if required parameters are non-NULL, unless the documentation says that a argument can be NULL we dont need to check for it: the user may just as well provide a pointer to invalid memory, which we can't protect from anyway.

Reply:
👍 Also moved this and related parsing methods into `rdkafka_sasl_oauthbearer.c`